### PR TITLE
Acquisition mode: scanned fluorescence-lifetime and optical path manager

### PR DESCRIPTION
**Pull request by amuskens**
Includes some additional testing files that may not need to be tracked later on.

### CODE REVIEW

File: src/odemis/acq/test/path_test.py [hunk 1]
```diff
@@ -48,11 +48,12 @@
 CONFIG_PATH = os.path.dirname(odemis.__file__) + "/../../install/linux/usr/share/odemis/"
 # Test for the different configurations
 SPARC_CONFIG = CONFIG_PATH + "sim/sparc-sim.odm.yaml"
-SECOM_CONFIG = CONFIG_PATH + "sim/secom2-sim.odm.yaml"
```
Comment (amuskens) [inline src/odemis/acq/test/path_test.py]: For some reason, the backend would not start without this change.

Comment (pieleric) [reply src/odemis/acq/test/path_test.py]: Well, as you can guess, that's not the proper way to fix it ;-) Let's have a look on your computer what's preventing the secom2 simulator from starting. Might need to install or configure something extra.

File: src/odemis/acq/stream/_sync.py [hunk 2]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+        return px_dt
+
+    def _setLightEmission(self, value):
+        # set all light emissions at once to a value
+        em = self._light.emissions.value
+        em = [value] * len(em)
+        self._light.emissions.value = em
+
+    def _runAcquisition(self, future=None):
+
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Please add here a long comment to explain how the acquisition works. (In particular, explain all this game with the time correlator, the scanner and the tc_scanner and why we don't care about the data returned by the time correlator).

File: src/odemis/acq/stream/_sync.py [hunk 3]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+        logging.debug("Starting job: acquisition")
+
+        try:
+            self._new_acquisition = True
+            self._acq_done.clear()
+
+            px_dt = self._prepareHardware()
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: Get nfr also from `_prepareHardware()`, and avoid saving it as an attribute.

File: src/odemis/acq/stream/_sync.py [hunk 4]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+
+            # Start the acquisition
+            self._pdetector.data.subscribe(self._onNewData)
+            self._new_acquisition = True
+
+            # For each frame
+            for i in range(self._nfr):
+                # premature cancellation occurred if the flag has already been set.
+                if self._acq_done.is_set():
+                    raise RuntimeError("Acquisition canceled!!!")
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: No need to worry too much if the acquisition is cancelled. Avoid the "!!!", and raise a CancelledError.

File: src/odemis/acq/stream/_sync.py [hunk 5]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+            self._time_correlator.data.unsubscribe(self._onAcqStop)
+
+            # turn off the light
+            self._setLightEmission(0)
+
+            # Finish
+            self._acq_done.set()
+
+        except Exception as e:
+            logging.exception(e)
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: You need to also stop the acquisition in the case of an error. Otherwise the hardware might end up in some limbo state, where some detectors are still acquiring, and some stopped due to the failure.

File: src/odemis/acq/stream/_sync.py [hunk 6]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+            self._pdetector.data.subscribe(self._onNewData)
+            self._new_acquisition = True
+
+            # For each frame
+            for i in range(self._nfr):
+                # premature cancellation occurred if the flag has already been set.
+                if self._acq_done.is_set():
+                    raise RuntimeError("Acquisition canceled!!!")
+
+                self._frame_done.clear()
```
Comment (pieleric) [inline src/odemis/acq/stream/_sync.py]: This is quite tricky. If the frame acquisition is fast enough, `frame_done` might have been set() before you arrive here. And you will end up missing it. Maybe what would work better is:
* _onNewData() puts in a queue.Queue the timestamp (or any other stupid object) at which each image has been received
* Here you wait for a new element in the queue, with a timeout.

File: src/odemis/acq/stream/_helper.py [hunk 7]
```diff
@@ -1015,3 +1015,126 @@ def result_as_da(timeout=None):
...
+                dt = data.metadata[model.MD_DWELL_TIME]
+            except KeyError:
+                dt = data.metadata.get(model.MD_EXP_TIME, self.emitter.dwellTime.value)
+                logging.warning("No dwell time metadata found in the ScannedTCSettings data, "
+                                "will use %f s", dt)
+
+            d /= dt
+            assert isinstance(d, numbers.Real), "%s is not a number" % d
+
+        elif dtyp != model.MD_DT_NORMAL:
```
[the remark this answers is not in the document]
Comment (amuskens) [reply src/odemis/acq/stream/_helper.py]: Where would this be indicated?

File: src/odemis/acq/stream/_sync.py [hunk 8]
```diff
@@ -2000,3 +2001,212 @@ def _runAcquisition(self, future):
...
+
+    def _onAcqStop(self, dataflow, data):
+        pass
+
+    def _onNewData(self, dataflow, data):
+
+        self._frame_done.set()
+
+        logging.debug("New data received.")
+        if self._new_acquisition:
```
[the remark this answers is not in the document]
Comment (amuskens) [reply src/odemis/acq/stream/_sync.py]: How will this affect the image update thread? Is there a possibility that the latest data might not get updated?

File: src/odemis/acq/test/path_test.py [hunk 9]
```diff
@@ -1021,5 +1021,91 @@ def test_set_path_stream(self):
...
+
+    def setUp(self):
+        if self.backend_was_running:
+            self.skipTest("Running backend found")
+
+    def test_set_path_stream(self):
+        self.optmngr.setPath("confocal").result()
+
+        self.optmngr.setPath("flim").result()
+        assert_pos_almost_equal(self.detsel.position.value, {"rx": 1.57}, atol=1e-3)
```
Comment (amuskens) [inline src/odemis/acq/test/path_test.py]: This value fails by the way. Apparently it expects 3.14.

Comment (pieleric) [reply src/odemis/acq/test/path_test.py]: Indeed... I even wrote it in the YAML file myself. Sorry for the confusion. And apparently I set to just 3.14 (no extra digit), which is probably simpler in some way.

File: src/odemis/acq/test/path_test.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/acq/test/path_test.py]: Still need to revert this line. `secom-sim` ought to work fine.

File: src/odemis/acq/test/flim_test.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    def test_flim_acq_simple(self):
+        logging.debug("Testing acquisition")
+
+        helper = stream.ScannedTCSettingsStream('Stream', self.detector, self.ex_light, self.lscanner,
+                                                self.sft, self.apd, self.tc_scanner)
+
+        remote = stream.ScannedRemoteTCStream("remote", helper)
+
+        # Configure values and test acquisition for several dwell times.
+        for dwellTime in (1e-6, 6e-3):
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: You need to test a couple longer dwell times. At least, you should try something quite longer than the maximum scanner dwell time (~ 2e-3 s). For these longer dwell times, you can put a repetition of (1,1).

File: src/odemis/acq/test/flim_test.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+            helper.dwellTime.value = dwellTime  # seconds
+            helper.repetition.value = (64, 64)
+            helper.roi.value = (0, 0, 0.1, 0.1)
+            helper.image.subscribe(self.validate_image)
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: If a subscriber, eg `self.validate_image`, fails, the test will still pass fine (because it's in a separate thread). That's why you preferably want to just store the image in an attribute and validate it in the main thread of the unit test.

File: src/odemis/acq/test/flim_test.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+            return
+        test.stop_backend()
+
+    def _on_image(self, im):
+        # Is called when an image is received in the live setting stream
+        logging.info("image: dim %s" , im.shape)
+        self._image = im
+
+    def validate_scan_sim(self, remote, repetition):
+        self.assertIsNotNone(remote.raw)
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: You can check so many more things here!
At least, check that the raw[0].shape is the same as repetition[::-1].
You can also check that some of most important metadata is present and correct. eg, the `MD_DWELL_TIME` should be the `helper.dwellTime.value`. (BTW, that's probably not the case yet, as it's not updated when you integrate the data). You should check that `MD_PIXEL_SIZE` is (almost) the same as `helper.pixelSize.value`. You could even check that `MD_POS` is correct wrt the `.roi` (but that's quite a bit more involved).

File: src/odemis/acq/test/flim_test.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        self.assertIsNotNone(remote.raw[0].metadata)
+        
+    def validate_image(self, img):
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: Similar checks should go here, as for the raw. Check that the data is 3D (ie, RGB) and uint8. Check that the metadata fits the expectations.

File: src/odemis/acq/test/flim_test.py (hunk 12 above)
Comment (amuskens) [reply src/odemis/acq/test/flim_test.py]: I made modifications to the checks, but somehow the values for pixel size and repetitions are not as expected. I will have to ask you about how this should work.

File: src/odemis/driver/rigol.py [hunk 14]
```diff
@@ -64,7 +64,7 @@ def __init__(self, name, role, host, port=DEFAULT_PORT, channel=1, limits=(0, 10
         self._phase_shift = 0
         self._dc_bias = self._amplitude_pp / 2 + lo
 
-        self._duty_cycle = 50 #%
```
Comment (pieleric) [inline src/odemis/driver/rigol.py]: This shouldn't be part of the commit. Actually the change is only needed for the testing, so it can be completely dropped.

File: src/odemis/acq/test/flim_test.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        helper.is_active.value = True
+
+        # move spot
+        time.sleep(8.0)
+        spots.roi.value = (0.1, 0.3, 0.1, 0.3)
+        time.sleep(2.0)
+        spots.roi.value = (0.5, 0.2, 0.5, 0.2)
+        time.sleep(2.0)
+
+        # move spot
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: Wrong comment => `Stop the streams`.

File: src/odemis/acq/test/flim_test.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        helper.image.unsubscribe(self._on_image)
+        helper.is_active.value = False
+        spots.is_active.value = False
+
+        self.assertIsNotNone(self._image, "No data received")
+        self.assertIsInstance(self._image, model.DataArray)
+        self.assertIsNotNone(self._image.metadata['Acquisition date'], "No metadata")
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: Instead of explicitly writing `'Acquisition date'`, write `model.MD_ACD_DATE`, it reduces the chances of typos.

File: src/odemis/acq/test/flim_test.py (hunk 12 above)
Comment (pieleric) [reply src/odemis/acq/test/flim_test.py]: It's good that there is a check then!
Get the pixelSize and dwell time as arguments too. BTW, you could also check MD_POS.
As you compute the accumulation of DataArrays of the right shape and (*in theory*) the right metadata, I'd expect that MD_POS, MD_PIXEL_SIZE and the shape should be fine out of the box. MD_DWELL_TIME needs to be updated explicitly though.

File: src/odemis/acq/test/flim_test.py [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+            self.assertTrue(any(em > 0) for em in self.ex_light.emissions.value)
+            self.assertGreater(self.ex_light.power.value, 0)
+            self.assertGreater(self.ex_light.period.value, 0)
+            f.result()  # wait for the result. This blocks
+            time.sleep(0.3)
+            self.validate_scan_sim(remote, rep, helper.dwellTime.value,
+                                   helper.time_correlator.getMetadata()[MD_PIXEL_SIZE])
+            time.sleep(0.3)
+
+        # try some longer dwelttimes, with a smaller resolution
```
Comment (pieleric) [inline src/odemis/acq/test/flim_test.py]: This block of code is not used anymore. Please drop it.
